### DISCUSSION TITLE
Scaling to 5-10B + some questions

### DISCUSSION THREAD

**Discussion by hadim** [milvus-io/milvus#4942] [Q&A and General discussion] [answered]
Hello,

We are considering using Milvus for an internal project dealing with 5 to 10 billion sparse binary vectors of size 256 (or 512) that we are planning to deploy to a Kubernetes cluster.

- The vectors are very sparse (<1% of 1 value), does Milvus apply some kind of compression to it and does it leverage this when computing distances such as Jaccard (using pop count for example)?
- Have you already scaled Milvus to 5-10 billion vectors?
- Can the indexing step be distributed using mishards (couldn't find the information in the doc, my bad if I missed it...).
- My understanding is that for now, Milvus can only deal with integer indices (looking at https://github.com/milvus-io/milvus/issues/4810). Ideally, we would use 27 characters indices from ASCII.
- Is that possible to deploy different types of vectors (binary-512-sparse-jaccard and float-256-dense-l2) in the same instance and specify which one to use when requesting a similarity search in the parameters?

Thank you and great work on Milvus, this looks like a very nice project!

**Comment by yhmo** [comment, marked as answer]
10 billion binary vectors. 1 billion = 1000M
If the dimension is 256, total data size is 10x1000Mx256/8 bytes=320000MB=320GB
If the dimension is 512, total data size is 640GB

This is a huge dataset, it is a big challenge to milvus.

> * The vectors are very sparse (<1% of 1 value), does Milvus apply some kind of compression to it and does it leverage this when computing distances such as Jaccard (using pop count for example)?

Milvus data management doesn't compress vector data even the binary vector is sparse. I think this is a valuable feature for huge dataset case.

> * Have you already scaled Milvus to 5-10 billion vectors?

For float vectors, the largest dataset we tested is 1B 128-dimensional vectors.
But for binary vectors, we didn't test so huge dataset on our machines. The largest dataset we ever tested:
```
50M binary vectors, dimension=512
nq=1, topk=100, no index(FLAT), Jaccard metric type
search performance 60ms on a 20+ cpu cores machine
```

> * Can the indexing step be distributed using mishards (couldn't find the information in the doc, my bad if I missed it...).

The mishards website on github: https://github.com/milvus-io/milvus/blob/1.x/shards/README.md
Mishards has only one write node, only the write node can process data change such as insert/index, that means mishards cannot distribute index task to different machines. But, for cpu version, milvus execute index task(each task is for one segment) one by one; for gpu version, milvus can distribute index task to multiple gpu devices(these gpu devices are installed on the machine with the write node is deployed).

> * My understanding is that for now, Milvus can only deal with integer indices (looking at #4810). Ideally, we would use 27 characters indices from ASCII.

Lot of users complain about the integer id, they want string id. We could not implement this in v1.x because there would be large scale architect change. The new milvus v2.0 will support this.
As I know, some users use a SQL service(mysql, postgresql, etc.) to store a mapping between string id and integer id, they firstly query from milvus to get some integer ids, then use these integer ids to perform a sql query to SQL service get string ids. 

> * Is that possible to deploy different types of vectors (binary-512-sparse-jaccard and float-256-dense-l2) in the same instance and specify which one to use when requesting a similarity search in the parameters?

You can read the milvus website to learn the basic concept of milvus: https://www.milvus.io/docs/v1.0.0/storage_concept.md
If you have two type of vectors, you can create two collections, the concept 'collection' is equal to the 'table' in SQL service. Each search request is to a single collection.

**Reply by hadim** [reply]
Thank you for the detailed answer. I guess index building not being distributed is kind of challenging for such a large dataset.

**Reply by hadim** [reply]
Do you know how long it took to index the dataset you talked about `50M binary vectors, dimension=512` ?

**Reply by yhmo** [reply]
> Thank you for the detailed answer. I guess index building not being distributed is kind of challenging for such a large dataset.

Yes, this is limitation of milvus v1.0
The milvus v2.0 is a real distributed system, it can distribute index tasks to multiple index nodes.

**Reply by yhmo** [reply]
> Do you know how long it took to index the dataset you talked about `50M binary vectors, dimension=512` ?

I don't know. We didn't build index for the 50M binary vectors, the test result is from brute-force search,

**Reply by hadim** [reply]
I tested building a collection of 1B 256-binary vectors (randomly generated so not sparse) with Milvus 1.0 and I am quite impressed by the performance of the insertion and indexing. Insertion took ~40min and indexing took ~60min (didn't go to the end and I still need to finish the prototype to look into search performance).

I am quite excited by Milvus 2.0 which seems to be distributed by design and compatible with string indices. Do you have an ETA for a first release? Will the upgrade between 1.x and 2.0 be smooth in your opinion?

**Reply by yhmo** [reply]
The 2.0 has different storage format, so the data cannot be compatible between 1.x and 2.0, but we will provide python tools to extract data from 1.x so that users can import to 2.0.
The 2.0 is still in testing/stable stage, hopefully the beta version can be released in 1 ~ 2 months.
